sdk/core/azure-core-amqp/src/amqp/management.cpp

```diff
@@ -249,5 +377,325 @@ namespace Azure { namespace Core { namespace Amqp { namespace _detail {
+
+    return Models::_internal::Messaging::DeliveryRejected(condition, description);
+  }
+
+  Models::AmqpValue ManagementClientImpl::OnMessageReceived(
+      _internal::MessageReceiver const&,
+      Models::AmqpMessage const& message)
+  {
+    if (message.ApplicationProperties.empty())
+    {
+      return IndicateError(
+          Models::_internal::AmqpErrorCondition::InternalError.ToString(),
+          "Received message does not have application properties.");
+    }
+    if (!message.Properties.CorrelationId.HasValue())
+    {
+      return IndicateError(
+          Models::_internal::AmqpErrorCondition::InternalError.ToString(),
+          "Received message correlation ID not found.");
+    }
+    else if (message.Properties.CorrelationId.Value().GetType() != Models::AmqpValueType::Ulong)
+    {
+      return IndicateError(
+          Models::_internal::AmqpErrorCondition::InternalError.ToString(),
+          "Received message correlation ID is not a ulong.");
+    }
+    uint64_t correlationId = message.Properties.CorrelationId.Value();
+
+    auto statusCodeMap = message.ApplicationProperties.find(m_options.ExpectedStatusCodeKeyName);
+    if (statusCodeMap == message.ApplicationProperties.end())
+    {
+      return IndicateError(
+          Models::_internal::AmqpErrorCondition::InternalError.ToString(),
+          "Received message does not have a " + m_options.ExpectedStatusCodeKeyName
+              + " status code key.");
+    }
+    else if (statusCodeMap->second.GetType() != Models::AmqpValueType::Int)
+    {
+      return IndicateError(
+          Models::_internal::AmqpErrorCondition::InternalError.ToString(),
+          "Received message " + m_options.ExpectedStatusCodeKeyName + " value is not an int.");
+    }
+    int32_t statusCode = statusCodeMap->second;
+
+    // If the message has a status description, remember it.
+    auto statusDescription
+        = message.ApplicationProperties.find(m_options.ExpectedStatusDescriptionKeyName);
+    std::string description;
+    if (statusDescription != message.ApplicationProperties.end())
+    {
+      if (statusDescription->second.GetType() != Models::AmqpValueType::String)
+      {
+        return IndicateError(
+            Models::_internal::AmqpErrorCondition::InternalError.ToString(),
+            "Received message " + m_options.ExpectedStatusDescriptionKeyName
+                + " value is not a string.");
+      }
+      description = static_cast<std::string>(statusDescription->second);
+    }
+
+    if (correlationId != m_expectedMessageId)
+    {
+      return IndicateError(
+          Models::_internal::AmqpErrorCondition::InternalError.ToString(),
+          "Received message correlation ID does not match request ID.");
+    }
+    if (!m_sendCompleted)
+    {
+      Log::Stream(Logger::Level::Error) << "Received message before send completed.";
+    }
+    if ((statusCode < 200) || (statusCode > 299))
```

Comment:
What dictates these status codes mean failed?

Reply:
I believe that the normative reference is [RFC 2616, section 6.1.1](https://www.rfc-editor.org/rfc/rfc2616#section-6.1.1).

The immediate reference is: https://www.oasis-open.org/committees/download.php/52425/amqp-man-v1%200-wd08.pdf, section 3.2.